src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSource.cs

```
@@ -70,6 +70,8 @@ public virtual async Task<T> GetAsync<T>(
             ILogger log,
             CancellationToken token)
         {
+            ThrowIfInsecureUri(request.Uri);
```

The current implementation throws an exception when an HTTP endpoint is accessed during any NuGet operation. I’m wondering if we should also surface an error earlier i.e. when locating the package source. At that point, we’ve already downloaded the index.json (either from the source or the HTTP cache) and populated the object model. If we detect that any of the resource endpoints use the HTTP scheme, we could fail fast and inform the user.

For example, we could say:

- "This package source contains only HTTP resource endpoints. NuGet doesn't allow accessing this package source."
- Or: "This package source contains a mix of HTTP and HTTPS resource endpoints. NuGet will block any operation that attempts to access an HTTP resource."

We can also include a note in the message indicating that customers can configure the `AllowInsecureConnections` setting if they want to enable access to HTTP resource endpoints.

This would help customers understand the issue earlier in the process and avoid restore attempts that are bound to fail.

---

That’s a great suggestion! However, we want to allow users to proceed with restore operations even if other resources, such as the search service, are using HTTP. Our goal is to be as permissive as possible and not block the entire package source simply because some endpoints are HTTP. 
Do you think that approach would be better?

---

>Our goal is to be as permissive as possible and not block the entire package source simply because some endpoints are HTTP.

I agree but in this scenario, we can say `This package source contains a mix of HTTP and HTTPS resource endpoints. NuGet will block any operation that attempts to access an HTTP resource. `

I am okay to address this comment in a follow-up.

---

https://github.com/NuGet/Home/issues/14420 tagging